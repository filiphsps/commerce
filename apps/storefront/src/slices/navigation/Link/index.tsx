import { linkResolver } from '@/utils/prismic';
import { cn } from '@/utils/tailwind';
import { asLink, type Content } from '@prismicio/client';
import { type SliceComponentProps } from '@prismicio/react';

import { LINK_ACTIVE_STYLES, LINK_STYLES } from '@/components/header/header-navigation';
import Link from '@/components/link';
import { PrismicText } from '@/components/typography/prismic-text';

export type LinkProps = {} & SliceComponentProps<Content.LinkSlice, { isHeader: boolean; pathname: string }>;
const LinkSlice = ({ slice, context: { isHeader = true, pathname } }: LinkProps) => {
    // Don't render the link slice as a standalone menu.
    if (!isHeader) return null;

    const link = slice.primary.href;
    const title = slice.primary.title;
    const variant = slice.variation;

    let target = asLink(link, { linkResolver });
    // Make sure the target ends with a slash.
    // TODO: This should be a utility function that wraps `asLink`.
    if (target && target.length > 1 && !target.endsWith('/')) {
        target = `${target}/`;
    }

    let active = target ? pathname.toLowerCase().endsWith(target.toLowerCase()) : false;
    // Handle homepage.
    if (active && target === '/') {
        active = pathname.split('/').length === 3;
    }

    let linkStyles = '';
    switch (variant as any) {
        case 'highlighted':
            linkStyles =
                cn(
                    LINK_STYLES,
                    'px-3 py-0 md:px-3 h-8',
                    active && LINK_ACTIVE_STYLES,
                    'text-extrabold rounded-lg bg-secondary-light text-secondary-foreground',
                    active && 'bg-primary text-primary-foreground'
                ) || '';
            break;
        default:
            linkStyles = cn(LINK_STYLES, active && LINK_ACTIVE_STYLES) || '';
            break;
    }

    return (
        <Link key={slice.id} href={target} className={linkStyles} data-active={active}>
            <PrismicText data={title} styled={false} bare={true} />
        </Link>
    );
};

export default LinkSlice;
