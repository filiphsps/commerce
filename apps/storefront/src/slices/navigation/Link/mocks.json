[
    {
        "__TYPE__": "SharedSliceContent",
        "variation": "default",
        "primary": {
            "title": {
                "__TYPE__": "StructuredTextContent",
                "value": [
                    {
                        "type": "paragraph",
                        "content": {
                            "text": "Ut aliqua id velit non ex exercitation laboris quis eu ut cillum."
                        }
                    }
                ]
            },
            "href": {
                "__TYPE__": "LinkContent",
                "value": {
                    "__TYPE__": "ExternalLink",
                    "url": "https://prismic.io"
                }
            }
        },
        "items": []
    },
    {
        "__TYPE__": "SharedSliceContent",
        "variation": "highlighted",
        "primary": {
            "title": {
                "__TYPE__": "StructuredTextContent",
                "value": [
                    {
                        "type": "paragraph",
                        "content": {
                            "text": "Aliquip cillum fugiat enim fugiat laborum quis excepteur sint. Sint nulla eiusmod laborum."
                        }
                    }
                ]
            },
            "href": {
                "__TYPE__": "LinkContent",
                "value": {
                    "__TYPE__": "ExternalLink",
                    "url": "https://slicemachine.dev"
                }
            }
        },
        "items": [
            {
                "__TYPE__": "GroupItemContent",
                "value": []
            }
        ]
    }
]
