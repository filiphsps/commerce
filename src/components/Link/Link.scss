@import '../../scss/app';

.Link {
    display: inline;
    cursor: pointer;
}
