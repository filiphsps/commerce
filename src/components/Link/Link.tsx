import NextLink from 'next/link';
import { useRouter } from 'next/router';
import React from 'react';

const Link = (props) => {
    const router = useRouter();

    return (
        <NextLink
            href={props.as || props.to}
            as={props.as ? props.to : undefined}
            locale={props.locale ? props.locale : router.locale}
        >
            <a
                {...props}
                className={`Link ${props.className}`}
                style={props.style}
            >
                {props.children}
            </a>
        </NextLink>
    );
};

Link.defaultProps = {
    to: '',
    className: ''
};

export default Link;
