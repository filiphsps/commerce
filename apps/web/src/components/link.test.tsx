import type { Locale } from '@/utils/locale';
import { render } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import Link from './link';

describe('components', () => {
    describe('Link', () => {
        beforeEach(() => {
            vi.spyOn(console, 'warn').mockImplementation(() => {});
            vi.spyOn(console, 'error').mockImplementation(() => {});
        });

        vi.mock('@/utils/build-config', () => ({
            BuildConfig: {
                domain: 'example.com',
                i18n: {
                    default: 'en-US'
                }
            }
        }));

        it('should render a link with the correct `href`', () => {
            const href = '/some/path';
            const { container } = render(<Link href={href} />);
            const link = container.querySelector('a');
            expect(link).not.toBeNull();
            expect(link?.getAttribute('href')).toBe(`/en-US${href}`);
        });

        it('should throw an error if `href` is not a string', () => {
            const href = { invalid: 'href' };
            expect(() => render(<Link href={href as any} />)).toThrowError();
        });

        it('should add the locale to the `href` if it is not already present', () => {
            const href = '/some/path';
            const locale: Locale = { locale: 'sv-SE', language: 'sv', country: 'SE' } as any;
            const { container } = render(<Link href={href} locale={locale} />);
            const link = container.querySelector('a');
            expect(link?.getAttribute('href')).toBe(`/${locale.code}${href}`);
            expect(link).toMatchSnapshot();
        });

        it('should not add the locale to the `href` if it is already present', () => {
            const href = '/sv-SE/some/path';
            const locale: Locale = { locale: 'sv-SE', language: 'sv', country: 'SE' } as any;
            const { container } = render(<Link href={href} locale={locale} />);
            const link = container.querySelector('a');
            expect(link?.getAttribute('href')).toBe(href);
            expect(link).toMatchSnapshot();
        });

        it('should remove the current domain from the `href`', () => {
            vi.spyOn(window, 'location', 'get').mockReturnValue({ host: 'example.com' } as any);
            window.location.host = 'example.com';
            const href = `https://example.com/some/path`;
            const { container } = render(
                <Link
                    shop={
                        {
                            domains: {
                                primary: 'example.com'
                            }
                        } as any
                    }
                    href={href}
                />
            );
            const link = container.querySelector('a');
            expect(link?.getAttribute('href')).toBe('/en-US/some/path');
            expect(link).toMatchSnapshot();
        });

        it('should remove double slashes from the `href`', () => {
            const href = '/some//path';
            const { container } = render(<Link href={href} />);
            const link = container.querySelector('a');
            expect(link?.getAttribute('href')).toBe('/en-US/some/path');
            expect(link).toMatchSnapshot();
        });

        it('should pass through all other props to the underlying link', () => {
            const href = '/en-US/some/path';
            const className = 'my-link';
            const { container } = render(<Link href={href} className={className} />);
            const link = container.querySelector('a');
            expect(link?.getAttribute('href')).toBe(href);
            expect(link?.getAttribute('class')).toBe(className);
        });

        it('should handle `href` using `tel:`, `mailto:`, etc protocols', () => {
            const href = 'mailto:hi@nordcom.io';
            const { container } = render(<Link href={href} />);
            const link = container.querySelector('a');
            expect(link?.getAttribute('href')).toBe(href);
            expect(link).toMatchSnapshot();
        });
    });
});
