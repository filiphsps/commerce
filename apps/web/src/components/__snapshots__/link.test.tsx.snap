// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components > Link > should add the locale to the \`href\` if it is not already present 1`] = `
<a
  href="/undefined/some/path"
/>
`;

exports[`components > Link > should handle \`href\` using \`tel:\`, \`mailto:\`, etc protocols 1`] = `
<a
  href="mailto:hi@nordcom.io"
/>
`;

exports[`components > Link > should not add the locale to the \`href\` if it is already present 1`] = `
<a
  href="/sv-SE/some/path"
/>
`;

exports[`components > Link > should remove double slashes from the \`href\` 1`] = `
<a
  href="/en-US/some/path"
/>
`;

exports[`components > Link > should remove the current domain from the \`href\` 1`] = `
<a
  href="/en-US/some/path"
  shop="[object Object]"
/>
`;
